### ISSUE TITLE
question about gmsh

### ISSUE THREAD

**Issue by AliceNeper** [TarletonGroup/CrystalPlasticity#62]
Hello Enrlalp，Thank you for your email reply. I still have some questions after reading your youtube videos.
1.The parameters in the gnd model 3 in your example file do not correspond to those in your excel. I think it is because you added several empty parameters after the output parameter,right
2.The python script I used to assign material properties can run under the C3D8 unit (under the polycrystalline material inp file you gave), but the result does not convergence under the grid type of C3D4 divided by gmsh(Because I am not familiar with gmsh, I can only divide C3D4 units) in the model I created using naper. May I ask why？
3.You mentioned in your youtube video that if the gnd model is used, the number of grids and the type of grids must be determined in the userinput file, but I did not find the option of cell type in your latest userinput file, whether to add it or not?
4.If I use gnd3 model to modify the options in the subroutine, can you tell me what to pay attention to.

**Comment by EralpDemir** [comment]
Hello,
1. There has been many updates since the initial version of the code. That is true. But check out the excel file under the documentation folder. That is the latest version.
2. The GND model is not available for C3D4 type element. You may generate the elements for C3D4 and then convert to C3D10 using abaqus. That may work. There is a generator file for Neper, if you follow the procedure that should work.
3. Currently, the element number and type determined automatically. There is a counter that does it in the very first cycle. Therefore, the very first cycle does not converge. 
4. Just setting gndmodel=3 should do it. It is best to use the default settings for the rest. 

If you need further help, you may share your inp file and I can look into the problem.

Best

**Comment by AliceNeper** [comment]
[cpfem.zip](https://github.com/user-attachments/files/16095805/cpfem.zip)
Hello,Enrlalp
  Thank you for your answer and I have modified the inp file. Unfortunately, I still could not get the final result for C3D10 unit, which did not converge in the calculation. So I used the C3D8 unit file that you used for neper partition, brought my data and used (hadeningmodel=3,gnd=3) into it to work out the result. Maybe I'm not very familiar with this model, and I think it's not normal to get a negative number. I also want to ask, if my material has a nanoprecipitated phase can use this subroutine? Thank you again for your help. These suggestions really mean a lot to me.

**Comment by EralpDemir** [comment]
Hi,
You may try generating C3D4 in Neper and convert to C3D10 using abaqus. The Neper2Abq works for C3D4 but it think I have not tested for C3D10.

GNDs can be positive or negative, so we keep their sign. There is also an output showing the sum of squares of the GND densities as a positive scalar value.

Yes you can use it for multiphase material. You must define a different materialID for each phase, define the properties in usermaterial.f file, which is not to be confused with the phase ID.

Best

**Comment by AliceNeper** [comment]
Dear Eralp:
I still have some problems when using the program and need to ask you for advice.
1.In terms of abaqus, the subprogram you provided specified that the units are μm and Mpa, but in the unit system of abaqus, Mpa corresponds to mm. I would like to know if I build a 1*1*1 cube, the unit is also μm, right?
2.When using hadeningmodel3, I found that if hadeningparameter (2) ==0, then the subroutine uses hadeningparameter 1,3,4,5,6 to calculate k2. But I changed the values in them, and none of the results changed.（The parameters I selected in your document are the same）
3.I established the model of equivalent circle diameter of 0.3 and 3, and calculated the stress-strain curve by using the FCC model of slip model=3, hadeningmodel=3 gnd=2. Although the strength of small grains is high, the yield strength does not change significantly, and the parameters are the same. I want to know the principle.
![1720178621661](https://github.com/TarletonGroup/CrystalPlasticity/assets/148926431/9f19dc22-7683-441b-98e5-8041f47b818c)

Looking forword to your reply.

Best wishes,
yuhong
